daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/PartialNextElementResolver.scala

```diff
@@ -302,7 +302,7 @@ class SeveralPossibilitiesForNextElement(
       }
 
     val res = optnextERD
-    if (isRequiredStreamingUnparserEvent && res.isEmpty) {
+    if (isRequiredStreamingUnparserEvent && res.isEmpty && !trd.isInstanceOf[ChoiceRuntimeData]) {
```

Comment:
I don't know enough about this code to know if this is the right fix, but here are some initial thoughts. Hoping @mbeckerle has a better idea.

I'm wondering if this issue isn't specific to choices? For example, say we had a shared group like which acts a little bit similar to the choice in this test:

```xml
  <group name="check">
    <sequence>
      <element name="not_valid" type="ex:str0" minOccurs="0"/>
    </choice>
  </group>
```

I guess this is different because in this case `isRequiredStreamingUnparseEvent` might be false because `not_valid` is optional which allows it to go up the stack?

Maybe that means in the case of this choice that has an empty branch `isRequiredStreamingUnparseEvent` should be false? I guess that isn't right though because the nackInfo/ackInfo that it's looking for next is required, it's just required outside the scope of this choice?

Another thought, should all choices skip this logic like you have now, or is this specific to choices that could have a branch with no element (e.g. empty sequence, all optional elements)?

Reply:
The only reason I think it might be choice specific is that yesterday while we were discussing this issue one of the things we tried was changing the choice inside the group to just a sequence and everything worked fine, so clearly wasn't hitting the same edge case.

As for the rest of your comment, I'll wait for Mike so weigh in as I'm really not sure what the best approach is.  Just saw a quick and easy way to fix it and saw that it worked for both shared groups (DAFFODIL-2615) and when the last branch of a choice in a group is optional (DAFFODIL-2562).

Reply:
It's not at all unlikely that your added predicate fixes an oversight in the original code. 
However, we do need a rationale for this, and we need to put a comment in that contains that rationale. 

This section of the code is checking whether we found a unique ERD corresponding to the incoming infoset element event, or not. 

The predicate you have enhanced is testing for 

1. the element is required to appear in the data stream (not optional element within the current group)
2. no ERD is found for it. (in this ERD resolver from this TRD on the trd stack)

You've added that it is also only an error if the current term is not a choice group.  If it is not an error, then the search will continue with the next resolver down the stack.  I think Steve is perhaps right that this is not about choices in general, but about choices with empty or all-optional branches. 

I'm inclined to think what is wrong is more likely to be that isRequiredStreamingUnparserEvent is true when it should be false.  That's my theory anyway. 

As for how to test this, ... Re: steve's thought: _Another thought, should all choices skip this logic like you have now, or is this specific to choices that could have a branch with no element (e.g. empty sequence, all optional elements)?_

We should create tests that try choices where there are branches which are not the first branch, and which contain: 
- nothing (empty sequence)
- syntax only (like a terminator, but on an otherwise empty sequence)
- optional content only (two cases: a nested choice with an empty sequence branch, and minOccurs=0)

We definitely also need a test where the shared group is a sequence with only optional content. 

Once we have those tests, I think we'll be in a better position to write a rationale and understand if we're missing a concept for choices, or if the problem is with isRequiredStreamingUnparserEvent, or ... something else entirely.

Reply:
Following up to my own comment. I think the issue is that the choice whose TRD is on top of stack is either Open or Closed based on analysis in the schema compiler. The isRequiredStreamingUnparserEvent should be false if the choice is Open. A choice is closed if every branch is closed. A sequence is closed if it contains a required element. I suspect there is some flaw in the Open/Closed logic leading to incorrect isRequiredStreamingUnparser event.

Reply:
I've spent some time digging into the Open/Closed choice branches and it seems like we are doing the right thing with the choice branches where the Choice doesn't close until all the branches are closed.  However, at the bottom of possibleSelfPlusNextLexicalSiblingStreamingUnparserElements we have the following match statement:

`      val res = thisItself match {                                                                                                                                              
        //                                                                                                                                              
        // check this case separately first to avoid evaluating                                                                                                                                              
        // followingLexicalSiblingStreamingUnparserElements unless                                                                                                                                              
        // we have to.                                                                                                                                              
        //                                                                                                                                              
        case Closed(pnes1) => thisItself // most common case - a required element.                                                                                                                                              
        case _ => {                                                                                                                                              
          val res: PossibleNextElements =                                                                                                                                              
            (thisItself, followingLexicalSiblingStreamingUnparserElements) match {                                                                                                                                              
              case (Open(pnes1), Closed(Nil)) => {                                                                                                                                              
                System.err.println("Open + Closed")                                                                                                                                              
                thisItself // case of Open(...) followed by end of complex element.                                                                                                                                              
              }                                                                                                                                              
              case (poss1, Closed(pnes2)) => {                                                                                                                                              
                System.err.println("Closed")                                                                                                                                              
                Closed((poss1.pnes ++ pnes2).distinct)                                                                                                                                              
              }                                                                                                                                              
              case (poss1, poss2) => {                                                                                                                                              
                System.err.println("Open")                                                                                                                                              
                Open((poss1.pnes ++ poss2.pnes).distinct)                                                                                                                                              
              }                                                                                                                                              
            }                                                                                                                                              
          res                                                                                                                                              
        }
`

You can ignore the print statements that I added.  For these choice we are hitting the second case and therefore returning a Closed list of possible elements, which results in isRequiredStreamingUnparserEvent being true and triggers the ErrorERD being created and we get a failure.  I have tried changing the 2nd case to the following:

`case (poss1, Closed(pnes2)) if (!this.isInstanceOf[ChoiceTermBase]) => { `

Which results in our use case hitting the 3rd case and gets past the error we are seeing.  Note, this does not fix the issue in DAFFODIL-2562 so if this is an acceptable solution, more work will need to be done to support that issue.

Reply:
@mbeckerle Could not get the code formatted properly in that comment, but does that seem like a reasonable fix?

Reply:
Easiest way to include code is to wrap it in three backticks, you can also put `scala` after the first set to get the right highlighting:

````
```scala
// scala code here
val foo = ....
```
````

Reply:
We just need a lot more information to know if this is sensible or not. 

How about print out the term and the value of possibleThisTermNextStreamingUnparserElementsDef each time it is computed. 

Then at runtime, let's display every pushTRD and popTRD, and each time we try to resolve a incoming event, let's print out the result of trying each partial next element resolver.